lib/elixir/lib/uri.ex

```diff
@@ -1021,6 +1021,113 @@ defmodule URI do
   def append_path(%URI{}, path) when is_binary(path) do
     raise ArgumentError, ~s|path must start with "/", got: #{inspect(path)}|
   end
+
+  @doc """
+  Normalizes the given `uri` following the rules defined by RFC 3986.
+
+  ## Examples
+
+      iex> URI.normalize(URI.parse("HtTp://example.com/")) |> URI.to_string()
+      "http://example.com/"
+
+      iex> URI.normalize(URI.parse("http://ExAmPlE.CoM/")) |> URI.to_string()
+      "http://example.com/"
+
+      iex> URI.normalize(URI.parse("http://example.com/%61")) |> URI.to_string()
+      "http://example.com/a"
+
+      iex> URI.normalize(URI.parse("http://example.com/%7Ejane")) |> URI.to_string()
+      "http://example.com/~jane"
+
+      iex> URI.normalize(URI.parse("http://example.com/?q=1%2f2")) |> URI.to_string()
+      "http://example.com/?q=1%2F2"
+
+      iex> URI.normalize(URI.parse("http://example.com/a/./b")) |> URI.to_string()
+      "http://example.com/a/b"
+
+      iex> URI.normalize(URI.parse("http://example.com/./a/b/../c/./")) |> URI.to_string()
+      "http://example.com/a/c/"
+
+      iex> URI.normalize(URI.parse("http://example.com")) |> URI.to_string()
+      "http://example.com/"
+
+      iex> URI.normalize(URI.parse("http://example.com:/")) |> URI.to_string()
+      "http://example.com/"
+
+      iex> URI.normalize(URI.parse("http://example.com:80/")) |> URI.to_string()
+      "http://example.com/"
+
+      iex> URI.normalize(URI.parse("https://example.com:443/")) |> URI.to_string()
+      "https://example.com/"
+
+  """
+  @doc since: "1.17.0"
+  @spec normalize(t()) :: t()
+  def normalize(%URI{} = uri) do
+    scheme = normalize_scheme(uri.scheme)
+
+    %URI{
+      scheme: scheme,
+      userinfo: normalize_userinfo(uri.userinfo),
+      host: normalize_host(uri.host),
+      port: normalize_port(uri.port, scheme),
+      path: normalize_path(uri.path),
+      query: normalize_query(uri.query),
+      fragment: normalize_fragment(uri.fragment)
+    }
+  end
+
+  defp normalize_scheme(nil), do: nil
+  defp normalize_scheme(scheme), do: String.downcase(scheme)
+
+  defp normalize_userinfo(nil), do: nil
+
+  defp normalize_userinfo(userinfo) do
+    case String.split(userinfo, ":", parts: 2) do
+      [user | []] ->
+        reencode(user)
+
+      [user | [password]] ->
+        encoded_user = reencode(user)
+        encoded_password = reencode(password)
+        encoded_user <> ":" <> encoded_password
+    end
+  end
+
+  defp normalize_host(nil), do: nil
+  defp normalize_host(host), do: String.downcase(host)
+
+  defp normalize_port(port, scheme) do
+    case scheme && default_port(scheme) do
+      ^port -> nil
+      _ -> port
+    end
+  end
+
+  defp normalize_path(nil), do: "/"
+
+  defp normalize_path(path) do
+    path
+    |> remove_dot_segments_from_path()
+    |> reencode()
+  end
+
+  defp normalize_query(nil), do: nil
+
+  defp normalize_query(query) do
+    query
+    |> query_decoder(:rfc3986)
+    |> encode_query(:rfc3986)
```

Comment:
What is the purpose of query reencoding?

Reply:
I believe it is necessary because it also encodes reserved characters that serve as delimiters in the query component: 

```
iex(25)> URI.parse("http://example.org?q=%26").query |> URI.query_decoder(:rfc3986) |> URI.encode_query(:rfc3986)
"q=%26"
iex(26)> URI.parse("http://example.org?q=%26").query |> URI.decode() |> URI.encode()
"q=&"
```

But when I read the RFC again, I start to doubt whether normalization should touch reserved characters:

```
   The purpose of reserved characters is to provide a set of delimiting
   characters that are distinguishable from other data within a URI.
   URIs that differ in the replacement of a reserved character with its
   corresponding percent-encoded octet are not equivalent.  Percent-
   encoding a reserved character, or decoding a percent-encoded octet
   that corresponds to a reserved character, will change how the URI is
   interpreted by most applications.  Thus, characters in the reserved
   set are protected from normalization and are therefore safe to be
   used by scheme-specific and producer-specific algorithms for
   delimiting data subcomponents within a URI.
```

Maybe I should check again what other implementations ([Erlang](https://www.erlang.org/doc/man/uri_string#normalize-1), [Boost](https://www.boost.org/doc/libs/1_84_0/libs/url/doc/html/url/urls/normalization.html), [Hyperlink](https://github.com/python-hyper/hyperlink)) do.

Reply:
The RFC says "These URIs should be normalized by decoding any percent-encoded octet that corresponds to an unreserved character", so it seems we want a subset of the decoding behaviour that converts only some percent escapes (and normalizes all other to upcase).

However, given Erlang implements this function, should we simply tell folks to use the Erlang function instead?

Reply:
> However, given Erlang implements this function, should we simply tell folks to use the Erlang function instead?

I believe that would make the most sense 😅

I found the uri_string Erlang module quite late, when I was already implementing URI.normalize/1. I believe mentioning somewhere in the documentation of the URI module that "Erlang's uri_string module provides additional functionality like for example URI normalization" or something like that could help people to find it earlier 🙂

Reply:
Let's go with the docs route then. Would you like to send a PR adding such mention? Perhaps also worth calling out normalization explicitly before URI.new and URI.parse.

Reply:
> Would you like to send a PR adding such mention?

Yes, I will send it tomorrow (1:30 AM here).

Reply:
See #13525